src/meta/processors/admin/ListClusterInfoProcessor.cpp

```diff
@@ -22,52 +22,98 @@ void ListClusterInfoProcessor::process(const cpp2::ListClusterInfoReq& req) {
-      handleErrorCode(nebula::cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
-      onFinished();
-      return;
+    // fill the dir info
+    if (info.role_ != meta::cpp2::HostRole::AGENT) {
+      auto dirKey = MetaKeyUtils::hostDirKey(addr.host, addr.port);
+      auto dirRet = doGet(dirKey);
+      if (!nebula::ok(dirRet)) {
+        LOG(ERROR) << folly::sformat("Get host {} dir info for {} failed: {}",
+                                     addr.toString(),
+                                     apache::thrift::util::enumNameSafe(info.role_),
+                                     apache::thrift::util::enumNameSafe(nebula::error(dirRet)));
+        handleErrorCode(nebula::error(dirRet));
+        onFinished();
+        return;
+      }
+      auto dir = MetaKeyUtils::parseHostDir(std::move(nebula::value(dirRet)));
+      service.set_dir(std::move(dir));
     }
 
-    storages.emplace_back(apache::thrift::FragileConstructor(), std::move(host), status.value());
-    iter->next();
+    if (hostServices.find(addr.host) == hostServices.end()) {
+      hostServices[addr.host] = std::vector<cpp2::ServiceInfo>();
+    }
+    hostServices[addr.host].push_back(service);
   }
 
-  auto* pm = store->partManager();
-  auto* mpm = dynamic_cast<nebula::kvstore::MemPartManager*>(pm);
-  if (mpm == nullptr) {
-    resp_.set_code(nebula::cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
+  // meta service
+  auto partRet = kvstore_->part(kDefaultSpaceId, kDefaultPartId);
+  if (!nebula::ok(partRet)) {
+    auto code = nebula::error(partRet);
+    LOG(ERROR) << "get meta part store failed, error: " << apache::thrift::util::enumNameSafe(code);
+    handleErrorCode(nebula::cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
     onFinished();
     return;
   }
-  auto& map = mpm->partsMap();
-  auto hosts = map[kDefaultSpaceId][kDefaultPartId].hosts_;
-  LOG(INFO) << "meta servers count: " << hosts.size();
-  resp_.set_meta_servers(std::move(hosts));
+  auto raftPeers = nebula::value(partRet)->peers();
+  for (auto& raftAddr : raftPeers) {
+    auto metaAddr = Utils::getStoreAddrFromRaftAddr(raftAddr);
+    cpp2::ServiceInfo service;
+    service.set_role(cpp2::HostRole::META);
+    service.set_addr(metaAddr);
+
+    if (hostServices.find(metaAddr.host) == hostServices.end()) {
+      hostServices[metaAddr.host] = std::vector<cpp2::ServiceInfo>();
+    }
+    hostServices[metaAddr.host].push_back(service);
+  }
+
+  // check: there should be only one agent in each host
+  for (const auto& [host, services] : hostServices) {
+    int agentCount = 0;
+    for (auto& s : services) {
+      if (s.get_role() == cpp2::HostRole::AGENT) {
+        agentCount++;
+      }
+    }
+    if (agentCount != 1) {
+      LOG(ERROR) << folly::sformat("There are {} agent count is host {}", agentCount, host);
+      handleErrorCode(nebula::cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
+      onFinished();
+      return;
+    }
+
+    if (services.size() <= 1) {
```

Comment:
Sorry, I don't quite understand why we check this condition? Would you explain more about it

Reply:
The log below could explain it?

Reply:
The log do confuse me... I mean why we return error if only agent has reported itself to meta on single host. That could really happen, e.g., all services has been stopped except agent. But that is not crucial to br?

Reply:
Because this function suppose to use when all service are right. But indeed I should not have this hypothesis, I'll change it.